### PR TITLE
LTP: Fix test case mkdir02, mkdir04 issue

### PR DESCRIPTION
**Pull request by shaikshavali1**
Test cases(mkdir02, mkdir04) is failed/skipped
because it exit improperly in middle of test. This is due
to single process environment. Test case is modified
to work in single process environment.

### CODE REVIEW

File: tests/ltp/patches/fix_mkdir_mkdir02.patch [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+ }
+ 
+@@ -67,6 +67,19 @@ static void setup(void)
+ 	struct passwd *pw;
+ 	struct stat buf;
+ 	pid_t pid;
++	int usr_secbits;
++
++	/*
++	 * if program started as a root user, avoid clearing of
```
Comment (hukoyu) [inline tests/ltp/patches/fix_mkdir_mkdir02.patch]: Stars looks ugly in the comment. Please fix. Better you can use multiple line // comments

File: tests/ltp/patches/fix_mkdir_mkdir02.patch [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
++	 * effective capabilities when changing from user root
++         * non-root user.
```
Comment (hukoyu) [inline tests/ltp/patches/fix_mkdir_mkdir02.patch]: I think you ment to non-root user

File: tests/ltp/patches/fix_mkdir_mkdir02.patch [hunk 3]
```diff
@@ -0,0 +1,74 @@
+	This change is necessary to run this test case in sgx-lkl
+	environment. Test case is modified to work in single
+	process environment by Removing exit code from child and 
+	wait code from parent. FORK_OR_VFORK return always 0 in
+	sgxlkl environment. Application in sgx-lkl environment
+	start with root user if this is the case Effective
```
Comment (hukoyu) [inline tests/ltp/patches/fix_mkdir_mkdir02.patch]: This change is necessary to run this test case in sgx-lkl
	environment. Test case is modified to work in single
	process environment by removing exit code from child and 
	wait code from parent. FORK_OR_VFORK always returns 0 in
	sgxlkl environment. Application in sgx-lkl environment
	starts with root user and some permissions lost when user
	changed from root to non-root. Hence, test case is updated
	to handle this scenario as well.

**Review by hukoyu — CHANGES_REQUESTED** [review]
Mostly feedback is about comments. Grammatical and style issues.

File: tests/ltp/patches/fix_mkdir_mkdir02.patch (hunk 1 above)
Comment (shaikshavali1) [reply tests/ltp/patches/fix_mkdir_mkdir02.patch]: Ok, I will use the single line comments. modified the code and update patch

File: tests/ltp/patches/fix_mkdir_mkdir02.patch (hunk 2 above)
Comment (shaikshavali1) [reply tests/ltp/patches/fix_mkdir_mkdir02.patch]: I will correct the comment

File: tests/ltp/patches/fix_mkdir_mkdir02.patch (hunk 3 above)
Comment (shaikshavali1) [reply tests/ltp/patches/fix_mkdir_mkdir02.patch]: I will update the patch description as per your comment.

**Comment by shaikshavali1** [comment]
@hukoyu 
I pushed new commit for mkdir04 test case as well. Could you please check the new commit "https://github.com/lsds/sgx-lkl/pull/249/commits/841e9dbf68bb83ce6f0667110cf17045212a2123".

File: tests/ltp/patches/fix_mkdir_mkdir02.patch (hunk 1 above)
Comment (shaikshavali1) [reply tests/ltp/patches/fix_mkdir_mkdir02.patch]: Done

File: tests/ltp/patches/fix_mkdir_mkdir02.patch (hunk 2 above)
Comment (shaikshavali1) [reply tests/ltp/patches/fix_mkdir_mkdir02.patch]: Done

File: tests/ltp/patches/fix_mkdir_mkdir02.patch (hunk 3 above)
Comment (shaikshavali1) [reply tests/ltp/patches/fix_mkdir_mkdir02.patch]: Done

**Comment by shaikshavali1** [comment]
@hukoyu , I addressed the review comments could you please check once.

**Review by hukoyu — APPROVED** [review]
LGTM. cc @letmaik

File: tests/ltp/patches/fix_mkdir_mkdir02.patch [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+ 	SAFE_RMDIR(TESTDIR2);
+ }
+ 
+@@ -67,6 +67,16 @@ static void setup(void)
+ 	struct passwd *pw;
+ 	struct stat buf;
+ 	pid_t pid;
++	int usr_secbits;
++
++	// set the flag "SECBIT_NO_SETUID_FIXUP" to retain the
```
Comment (letmaik) [inline tests/ltp/patches/fix_mkdir_mkdir02.patch:44]: It's not clear why this patch is needed to make mkdir work. Could you add more details?
Also, please use `if(getuid() == 0)` since your intention is not comparing a boolean.

File: tests/ltp/patches/fix_mkdir_mkdir02.patch [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
++		prctl(PR_SET_SECUREBITS, usr_secbits, 0, 0, 0);
++	}
+ 
+ 	pw = SAFE_GETPWNAM("nobody");
+ 	nobody_uid = pw->pw_uid;
+@@ -74,7 +84,7 @@ static void setup(void)
+ 	pw = SAFE_GETPWNAM("bin");
+ 	bin_uid = pw->pw_uid;
+ 	bin_gid = pw->pw_gid;
+-
```
Comment (letmaik) [inline tests/ltp/patches/fix_mkdir_mkdir02.patch]: Remove formatting change in patch.

File: tests/ltp/patches/fix_mkdir_mkdir02.patch [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
++#include <sys/prctl.h>
++#include <linux/securebits.h>
+ 
+ #define TESTDIR1	"testdir1"
+ #define TESTDIR2	"testdir1/testdir2"
+@@ -54,10 +56,8 @@ static void verify_mkdir(void)
+ 		if (!fail)
+ 			tst_res(TPASS, "New dir inherited GID and S_ISGID");
+ 
+-		exit(0);
```
Comment (letmaik) [inline tests/ltp/patches/fix_mkdir_mkdir02.patch:31]: Why was this removed?

File: tests/ltp/patches/fix_mkdir_mkdir02.patch [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+-
++	
+ 	umask(0);
+ 
+ 	pid = SAFE_FORK();
+@@ -84,10 +94,7 @@ static void setup(void)
+ 		SAFE_MKDIR(TESTDIR1, 0777);
+ 		SAFE_STAT(TESTDIR1, &buf);
+ 		SAFE_CHMOD(TESTDIR1, buf.st_mode | S_ISGID);
+-		exit(0);
```
Comment (letmaik) [inline tests/ltp/patches/fix_mkdir_mkdir02.patch:59]: Same as above.

File: tests/ltp/patches/fix_mkdir_mkdir02.patch [hunk 8]
```diff
@@ -0,0 +1,61 @@
+diff --git a/testcases/kernel/syscalls/mkdir/mkdir02.c b/testcases/kernel/syscalls/mkdir/mkdir02.c
```
Comment (letmaik) [inline tests/ltp/patches/fix_mkdir_mkdir02.patch:14]: Summary description of patch is missing. See other patches.

**Review by letmaik — CHANGES_REQUESTED** [review]

File: tests/ltp/patches/fix_mkdir_mkdir02.patch (hunk 8 above)
Comment (shaikshavali1) [reply tests/ltp/patches/fix_mkdir_mkdir02.patch:14]: Done

File: tests/ltp/patches/fix_mkdir_mkdir02.patch (hunk 5 above)
Comment (shaikshavali1) [reply tests/ltp/patches/fix_mkdir_mkdir02.patch]: Done

File: tests/ltp/patches/fix_mkdir_mkdir02.patch (hunk 6 above)
Comment (shaikshavali1) [reply tests/ltp/patches/fix_mkdir_mkdir02.patch:31]: FORK_OR_VFORK in LTP returns always 0 in sgxlkl environment because sgx-lkl support single process. Original test case create a child  and exit from child. But in sgxlkl this will cause exit from the test program before test completion.  Updated the details in patch description.

File: tests/ltp/patches/fix_mkdir_mkdir02.patch (hunk 7 above)
Comment (shaikshavali1) [reply tests/ltp/patches/fix_mkdir_mkdir02.patch:59]: Updated the details in patch description.

File: tests/ltp/patches/fix_mkdir_mkdir02.patch (hunk 4 above)
Comment (shaikshavali1) [reply tests/ltp/patches/fix_mkdir_mkdir02.patch:44]: Modified the expression in if condition. Details about the patch is added in patch description.

**Comment by shaikshavali1** [comment]
@letmaik, 
Hi Maik,
I addressed your review comments. Could you please check and approve.

File: tests/ltp/patches/fix_mkdir_mkdir02.patch [hunk 9]
```diff
@@ -0,0 +1,61 @@
+This change is necessary to run this test case in sgx-lkl
+environment. Test case is modified to work in single
+process environment by removing exit code from child and
```
Comment (letmaik) [inline tests/ltp/patches/fix_mkdir_mkdir02.patch]: You're not removing an exit code, you're removing a call to `exit()`. What is a "wait code"? I couldn't see that in the patch.

Comment (shaikshavali1) [reply tests/ltp/patches/fix_mkdir_mkdir02.patch]: @letmaik, The 'tst_reap_children()' function makes the process wait for all of its
children.

Comment (letmaik) [reply tests/ltp/patches/fix_mkdir_mkdir02.patch]: @hukoyu Could you please help rephrase the description to make it more clear? Thanks!

Comment (hukoyu) [reply tests/ltp/patches/fix_mkdir_mkdir02.patch]: In the original test a child process is used, and parent process waits for child process to finish, and child process calls exit to return to parent process. The test case is modified to work in sgx-lkl single
process environment by removing codes related to exit from child and
wait on parent. FORK_OR_VFORK always returns 0 in sgxlkl environment. 
In addition, application in sgx-lkl environment starts with root user and 
admin permissions lost when user changed from root to non-root user. 
Test case is updated to handle this scenario as well.

Comment (hukoyu) [reply tests/ltp/patches/fix_mkdir_mkdir02.patch]: @shaikshavali1 I reprhased the description. Please fix the formatting and length of lines.

Comment (shaikshavali1) [reply tests/ltp/patches/fix_mkdir_mkdir02.patch]: @hukoyu,  modified the description as suggested.

Comment (shaikshavali1) [reply tests/ltp/patches/fix_mkdir_mkdir02.patch]: @letmaik ,
Hi Mike,
Could you please re-run the failed jobs in Azure pipe line?

In Azure pipeline build, below mentioned two test cases were failed before starting the test application in enclave. These failure are not related to my commit. 

languages-java-read_file-(debug)-(run-hw)-(8-ethreads) 
basic-global_vars_test-(debug)-(run-sw)-(8-ethreads)

**Review by hukoyu — APPROVED** [review]
LGTM cc @letmaik

**Review by letmaik — APPROVED** [review]
